apps/dotcom/sync-worker/src/UserDataSyncer.ts

```diff
@@ -370,47 +399,99 @@ export class UserDataSyncer {
 			this.reboot()
 			return
 		}
+
 		this.state.lastSequenceNumber++
 
-		if (event.type === 'mutation_commit') {
-			this.commitMutations(event.mutationNumber)
+		// this event type only exists to make us check the sequence id + number
+		// so its job is done.
+		if (event.type === 'maybe_force_reboot') return
+
+		// in the rare case that the replicator was behind us when we called registerUser,
+		// ignore any events that came in from before we got our initial data.
+		if (event.lsn < this.store.getCommittedData()!.lsn) {
+			this.log.debug('ignoring old event', event.lsn, '<', this.store.getCommittedData()!.lsn)
 			return
 		}
 
-		if (this.state.type === 'connected') {
-			assert(event.type === 'row_update', `event type should be row_update got ${event.type}`)
-			this.handleRowUpdateEvent(event)
-			return
+		transact(() => {
+			for (const ev of event.changes) {
+				if (ev.type === 'mutation_commit') {
+					this.commitMutations(ev.mutationNumber)
+					continue
+				}
+
+				assert(ev.type === 'row_update', `event type should be row_update got ${event.type}`)
+				this.handleRowUpdateEvent(ev)
+			}
+
+			this.store.commitLsn(event.lsn)
+		})
+
+		// make sure we have all the files we need
```

Comment:
When would these two cases happen?

Also, do we want to return in the if? feels like we'd want to add all of the missing files?

Reply:
1. for adding a missing guest file record, this would happen if the user adds a file_state but something went wrong when adding the guest file record after the mutation was applied. unlikely but better safe than sorry I reckon.

2. for deleting a guest file record that is no longer needed, this currently happens mainly if a shared file is set to private (we delete the file_state for guests in that case)

and no that return in the if shouldn't be there!

Reply:
Where do we delete the `file_state`, must have missed that?

Reply:
https://github.com/tldraw/tldraw/blob/facfa5d1689b7722035e645627b63b5bcd78d011/apps/dotcom/zero-cache/migrations/000_seed.sql#L59-L66

Reply:
Thanks, completely forgot about this 😄